Decouple scheduled task execution from task scheduling.

chrisvest opened the pull request:
By making the scheduled executor service send the tasks to be executed in their respective pools, instead of the scheduling pool, and instead use the scheduling pool exclusively to trigger said tasks.

This also includes a handful of refactorings around the job scheduling infrastructure. And all threads started by the `CentralJobScheduler` now run within `ThreadGroups` derived from their scheduling `Group`s.

chrisvest committed:
Make the JobScheduler delegate to the global pool for actually executing scheduled tasks, and instead use the scheduled pool exclusively to trigger said tasks.

chrisvest committed:
Simplify page cache warmer profile scheduling.

chrisvest committed:
Make the Neo4jJobScheduler assign a ThreadGroup for every group, and run each in their own thread pool.

chrisvest committed:
Strengthen error handling of recurring tasks and make causal clustering tests pass.

chrisvest committed:
Checkstyle fixes.

chrisvest committed:
Remove the concept of scheduling metadata.

chrisvest committed:
Move the Neo4jJobScheduler to a dedicated scheduler package, and rename it to CentralJobScheduler.

chrisvest committed:
Extract inner classes from the CentralJobScheduler.

chrisvest committed:
Make DefaultConversationSPITest pass.

chrisvest committed:
Create a specialised time based task scheduler.

Instead of using the ScheduledExecutorService, this new scheduler will ensure that tasks are both started in their respective target
thread pools, run with their given ThreadGroup, and do not overlap in time in the sense that one run of a recurring task will not
start until a prior run has finished. Overdue and delayed tasks are handled as well.

chrisvest committed:
Some additional TimeBasedTaskScheduler changes that improve timeliness, reaction time to new tasks submitions, and shuts the CentralJobScheduler down in the right order.

klaren commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/PageCacheWarmerKernelExtension.java:
```diff
@@ -126,13 +126,7 @@ private void tryStartProfile()
     {
         if ( profilingStarted.compareAndSet( false, true ) )
         {
-            // At this point, we are currently executing inside a *scheduled* executor thread. There are, at this time
```
:tada:

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/CentralJobScheduler.java:
```diff
@@ -0,0 +1,187 @@
+    private static final AtomicInteger INSTANCE_COUNTER = new AtomicInteger();
+    private static final Group SCHEDULER_GROUP = new Group( "Scheduler" );
+
+    private TimeBasedTaskScheduler scheduler;
```
`final`, same for `schedulerThread`

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/ThreadPoolManager.java:
```diff
@@ -0,0 +1,58 @@
+    {
+        pools.forEach( ( group, pool ) -> pool.cancelAllJobs() );
+        pools.forEach( ( group, pool ) -> pool.shutDown() );
+        return pools.values().stream().reduce( null,
```
I would prefer to have a simpler map-reduce instead:
```java
return pools.values().stream().map( ThreadPool::getShutdownException ).reduce( null, Exceptions::chain );
```

chrisvest committed:
Make the TimeBasedTasksSchedulerTest not flaky.

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:
```diff
@@ -0,0 +1,182 @@
+import org.neo4j.scheduler.JobScheduler.JobHandle;
+import org.neo4j.time.SystemNanoClock;
+
+final class TimeBasedTaskScheduler implements Runnable
```
This class seems overly complicated for what needs to be done. A simple blocking priority queue would surely suffice.

chrisvest replied:
You mean a DelayQueue? I don't think it's that complicated for balancing reaction time and proper shut down.

chrisvest replied:
I side-stepped the concurrency questions of the _blocking_ priority queue, and kept the existing concurrency scheme but switched to an ordinary priority queue.

klaren replied:
I was referring to the inbox and queue split as well. Can't submit just add it to the queue directly?

If we're having trouble with contention on the submit method, I'm more than sure we don't need to worry about timeliness.

chrisvest replied:
I like that they are split. They do different things as it is now. And one has to deal with multi-threading, while the other is single threaded and carefree.

klaren replied:
I don't see how multiple threads submitting and one thread polling would change that. The timekeeper thread will only work on the first object in the queue each time.

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/ScheduledTask.java:
```diff
@@ -0,0 +1,135 @@
+import org.neo4j.scheduler.JobScheduler.CancelListener;
+import org.neo4j.scheduler.JobScheduler.JobHandle;
+
+final class ScheduledTask implements JobHandle
```
A `javadoc` that documents the states and transitions would be useful.

klaren reviewed (changes requested):
This is a move in the right direction, nice to get some structure around the scheduler.

Since it feels like my requested changes will change a lot of the code, let's start with these so I don't review code that will go away anyway.

Was there any particular reason for not using `ScheduledThreadPoolExecutor` as a dispatcher for the other thread pools? Exactly the way you solved it for the profiler before.

chrisvest commented:
> Was there any particular reason for not using ScheduledThreadPoolExecutor as a dispatcher for the other thread pools?

Yes, I tried this at first, but it became very unwieldy to trace cancellations and ExecutionExceptions back and forth, not to mention also handling overdue tasks and avoiding overlapping execution. The dedicated scheduler thread is definitely simpler.

chrisvest committed:
Address some review comments.

chrisvest committed:
Make the TimeBasedTaskScheduler use a PriorityQueue for scheduling, instead of a sorted linked list.

This simplifies the code a bit, and should also have better performance charactaristics when there is a large number of scheduled tasks.

chrisvest commented:
@klaren I believe I've address all of your comments.

chrisvest committed:
Simplify the TimeBasedTaskScheduler.

This also fixes a bug where waitTermination would not unblock if a task was cancelled before it got to run for the first time.
This also removes a spinning behaviour of the scheduler, when recurring tasks ended up overdue by not completing in time for their next schedule.

chrisvest commented:
@klaren Take another look.

chrisvest committed:
Remove DebugUtil.trackTest and its last usage. This fixes #10722

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:
```diff
@@ -0,0 +1,108 @@
+    {
+        this.clock = clock;
+        this.pools = pools;
+        delayedTasks = new PriorityBlockingQueue<>( 42, DEADLINE_COMPARATOR );
```
While 42 might be the answer to everything, somehow I doubt that includes optimal initial capacity for a blocking priority queue. But I don't what the answer to that would be...

chrisvest replied:
The queue grows automatically, and you have to choose something in order to provide a comparator, so here we are.

klaren reviewed (approved):
I'm happy with the current state! One magic constant to ponder upon before merging.
